proposed/swift-interop.md

```diff
@@ -0,0 +1,183 @@
+
+For reference, explicitly declaring a function with the Swift or SwiftAsync calling conventions in Clang requires the "context" argument, the value that goes in the "self" register, as the last parameter or the penultimate parameter followed by the error parameter.
+
+The self register, like the error register and the async context register discussed later, is always a pointer-sized value. As a result, if we introduce any special intrinsic types for the calling convention, we don't need to make the type generic as we can always use a `void*` to represent the value at the lowest level.
+
+We have rejected the following designs for the reasons mentioned below:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+
+Rejected Option 1 seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables.
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use a special type like `SwiftError` on a `ref` or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError*` to indicate the error parameter;
+
+Both options characteristics: They both express that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. The first option represents the "error return" as a managed `ref` or `out` parameter. Like with `SwiftSelf`, we would throw an `InvalidProgramException` for a signature with multiple `SwiftError` parameters.
+
+Option 2 provides an alternative to Option 2 as `ref` and `out` parameters are not supported in `UnmanagedCallersOnly` methods. Option 1 likely provides more possible JIT optimizations as managed pointers are more easily reasoned about in some of our JIT/AOT intermediate representations, but supporting it in all of our goal scenarios (P/Invoke, UnmanagedCallersOnly, function pointers) would require some minor language changes to allow the managed pointers in the `UnmanagedCallersOnly` signature for these special types. Option 2 (`SwiftError*`) does not require any language changes.
```

Comment:
> Option 1 likely provides more possible JIT optimizations as managed pointers are more easily reasoned about in some of our JIT/AOT intermediate representations

This may be true in general. For this case, the JIT should be able to see all places that the address flown into and optimize it well. For example, JIT is able to optimize this fine:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunykGArgDtc2AGYwGASyEYGGGLgwAKAJRcenajx3TZDBAwC8DcgG4NuvRgBUDAA7GGAMgQXtVpgHYGN++50AXxpAoA=